### PR TITLE
Fix for BUG: multi-index excel header fails if all numeric

### LINKED ISSUE #11408: Pandas exporting to Excel (xls, xlsx) with multilevel columns
Hello all,

I want collect few tables into the one with pandas. I use the code presented below:

```python
import pandas as pd
import itertools
import types

df = None
for frame in ['x', 'y']:
    df_ = pd.read_excel(u'%s.xlsx' % frame)
    df_ = df_.set_index([u'time'])

    parameters = list(df_.columns)
    tuples = []
    for tup in itertools.product([frame,], parameters):
        tuples.append(tup)

    columns = pd.MultiIndex.from_tuples(tuples, names=[u'Frames',u'Parameters'])

    df_new = pd.DataFrame(columns=columns, index=df_.index)
    for par in parameters:
        df_new[frame, par] = df_[par]
    del df_

    if df is None:
        df = df_new
    else:
        df = pd.concat([df, df_new], axis=1)

df.to_excel('merged_xlsx.xlsx')
df.to_excel('merged_xls.xls')
```

Source data is ``x.xlsx``
x

and ``y.xlsx`` files.
y

XLS engine works well (``merged_xls.xls``):
merged_xls

But something is wrong (cells merging) with XLSX engine (``merged_xlsx.xlsx``):
merged_xlsx

Manual cells unmerging works in Excel:
merged_xlsx_unmerged

Is it bug in XLSX pandas engine (openpyxl)? Or what is wrong in my code?

Versions: ``python-2.7.10``, ``pandas-0.17.0``, ``openpyxl-2.3.0``.

P.S. This issue is copy of [my question on stackoverflow](http://stackoverflow.com/questions/33274716/pandas-exporting-to-excel-xls-xlsx-with-multilevel-columns). It was suggested as bug and was adviced to post here.

### PR DESCRIPTION
**Pull request by Dr-Irv**
replaces #11317
closes #11408

This includes updates to 3 Excel files, plus a test in test_excel.py,
plus the fix in parsers.py

### CODE REVIEW

**Comment by jreback** [comment]
ok, pls add a whatsnew note in bug fixes (reference this PR number)

**Comment by jreback** [comment]
ideally I'd like a round-trip test, e.g. take a csv write to excel, read it back and compare

File: pandas/io/tests/test_excel.py [hunk 1]
```diff
@@ -557,6 +557,12 @@ def test_read_excel_multiindex(self):
         actual = read_excel(mi_file, 'mi_column_name', header=[0,1], index_col=0)
         tm.assert_frame_equal(actual, expected)
 
+        # Issue #11317
+        expected.columns = mi.set_levels(['1','2'],level=1).set_names(['c1', 'c2'])
```
Comment (jreback) [inline pandas/io/tests/test_excel.py]: show the construction of expected here (I know it is probably above), but nice to have it actually here

Comment (Dr-Irv) [inline pandas/io/tests/test_excel.py]: I followed the style in that particular function. It keeps modifying the stuff above, as multiple sheets are read in and tested for different variations.

**Comment by Dr-Irv** [comment]
The check failed for Python 2.7, so I have committed a change for that to parsers.py and also updated what's new for v0.17.1.

**Comment by jreback** [comment]
pls squash. ping when green.

**Comment by jreback** [comment]
sorry, meant to have you add a round-trip test as well.

**Comment by Dr-Irv** [comment]
I've added a round trip test. Discovered an issue where the Excel file isn't properly formatted when you have a MultiIndex column and merge_cells=false.  Not sure how to test for that.  Submitted code and assuming the tests will kick in

File: pandas/io/tests/test_excel.py [hunk 2]
```diff
@@ -1083,6 +1089,31 @@ def test_to_excel_multiindex(self):
...
+        new_cols_index = MultiIndex.from_tuples([(40,1),(40,2),(50,1),(50,2)])
+        frame.columns = new_cols_index
+
+        with ensure_clean(self.ext) as path:
+             # round trip
+            frame.to_excel(path, 'test1', merge_cells=self.merge_cells)
+            reader = ExcelFile(path)
+            df = read_excel(reader, 'test1', header=[0,1],
+                             index_col=[0, 1],
+                              parse_dates=False)
```
Comment (jreback) [inline pandas/io/tests/test_excel.py]: can you fix formatting

File: pandas/io/tests/test_excel.py [hunk 3]
```diff
@@ -1083,6 +1089,31 @@ def test_to_excel_multiindex(self):
...
+            tm.assert_frame_equal(frame, df)
+            self.assertEqual(frame.index.names, df.index.names)
```
Comment (jreback) [inline pandas/io/tests/test_excel.py]: ``assert_frame_equal`` asserts the index names are equal

**Comment by jreback** [comment]
@Dr-Irv minor fix squash

@chris-b1 can you have a look

**Comment by chris-b1** [comment]
Seems fine to me.  It's not introduced by this PR, but one inconsistency I just noticed is that `read_csv` seems to always converts numeric columns to strings (or maybe never converts to numeric in the first place), where `read_excel` will leave as numeric types.

File: pandas/io/tests/test_excel.py (hunk 3 above)
Comment (Dr-Irv) [inline pandas/io/tests/test_excel.py]: I deleted that, but also deleted it from the test above, which is where I copied the code from.

File: pandas/io/tests/test_excel.py (hunk 2 above)
Comment (Dr-Irv) [inline pandas/io/tests/test_excel.py]: Done.  See next commit

**Comment by jreback** [comment]
@chris-b1 can u file an issue about your last
and show an example

File: pandas/io/tests/test_excel.py [hunk 4]
```diff
@@ -1082,7 +1088,31 @@ def test_to_excel_multiindex(self):
...
+    def test_to_excel_multiindex_cols(self):
+        _skip_if_no_xlrd()
+        if not self.merge_cells:
+            raise nose.SkipTest('Skip tests MI on cols with no merging.')
+
+
+        frame = self.frame
+        arrays = np.arange(len(frame.index) * 2).reshape(2, -1)
+        new_index = MultiIndex.from_arrays(arrays,
+                                           names=['first', 'second'])
```
Comment (jreback) [inline pandas/io/tests/test_excel.py]: are the provies xls* files still used?

**Comment by jreback** [comment]
needs a squash

File: pandas/io/tests/test_excel.py (hunk 4 above)
Comment (Dr-Irv) [inline pandas/io/tests/test_excel.py]: Yes they are.  I also added some to illustrate the bug.

**Comment by Dr-Irv** [comment]
@jreback I think I have done everything that is needed. (Note - this is my first time doing any bug fixing with pandas or any open source project for that matter!) There is one issue and I don't know how I can debug it. In the test test_to_excel_multiindex_cols, if merge_cells is False, it fails on Python 2.7 and Python 3.5 according to the Travis CI testing. What is happening is that the multiindex is made from tuples [(40,1),(40,2),(50,1),(50,2)].  When merge_cells is False, the created column names are [ "40.1", "40.2", "50.1", "50.2" ].  On Python 2.7 and 3.5, it is reading those as floats rather than strings.  I did set up a Python 2.7 environment for debugging, but could not replicate. I could not set up a Python 3.5 environment for debugging (This is all on Windows).  So I have put something in there that makes the tests pass, but someone who knows how to do cross-platform debugging should look at this.

Thanks for your patience in letting me figure this out.

**Comment by jreback** [comment]
http://pandas.pydata.org/pandas-docs/stable/contributing.html#creating-a-development-environment for creating an environment on windows is quite easy (for any python).

As fair as the issue, maybe @chris-b1 has a thought?

**Comment by chris-b1** [comment]
That is along the same lines as  #11357.

Basically the type inference would need to happen after the splitting,
probably worth checking how csv handles it?

http://pandas.pydata.org/pandas-docs/stable/contributing.html#creating-a-development-environment
for creating an environment on windows is quite easy (for any python).

As fair as the issue, maybe @chris-b1 <https://github.com/chris-b1> has a
thought?

—
Reply to this email directly or view it on GitHub
<https://github.com/pydata/pandas/pull/11328#issuecomment-149354108>.

**Comment by Dr-Irv** [comment]
@jreback I'm unsure how to figure this out.  I built a Python 2.7 environment on Linux, and can't replicate the bug.  But the travis-ci build failed.  Here is the part that is confusing.  One of the Python 2.7 runs succeeded and the other one failed.  See https://travis-ci.org/Dr-Irv/pandas/builds/86225791 . Job 5.2 is Python 2.7 and succeeded.  Job 5.3 is Python 2.7 and failed. How can I replicate the failure that corresponds to Job 5.3?  When I run the nosetests on test_excel.py on Python 2.7, it passes (with my next-to-last version of test_excel.py).

**Comment by chris-b1** [comment]
Do you have all the dependencies (xlrd, xlsxwriter, openpyxl, xlst) installed locally? Might be skipping the failing test.

**Comment by Dr-Irv** [comment]
@chris-b1 Yes.  I even explicitly ran the test that failed (according to the log) and it passes (no skip message). I turned on the "-v" flag and see that it is run.  Also, it doesn't explain why in the Travis-CI tests, one 2.7 run succeeded and the other 2.7 run failed.

**Comment by chris-b1** [comment]
It looks like the passing 2.7 is just the "slow' tests.  Try installing the specific version of openpyxl in travis `openpyxl=1.6.2`  

https://github.com/pydata/pandas/blob/master/ci/requirements-2.7_LOCALE.run

**Comment by Dr-Irv** [comment]
@chris-b1 That was what I needed.  I can reproduce so I will now work on it.

**Comment by jreback** [comment]
@Dr-Irv yep this is why conda works well for this, I just create a separate environment with the contents of a particular travis run - then easy to repro

**Comment by Dr-Irv** [comment]
@chris-b1 @jreback So with Chris's hint, I was able to get it to fail on Linux. But on Windows, ci/requirements-2.7_LOCALE.run is not working with conda.  I'm trying to hack it to get it to work. Question for either of you - how do you tell from the Travis-CI runs which ci/requirements* file corresponds to which run?

**Comment by jreback** [comment]
in ``.travis.yml``, their is a JOB_NAME tag which corresponds to the run

LOCALE should work (though setting the LOCALE itself is tricky), see ``ci/script.py``

**Comment by Dr-Irv** [comment]
@jreback @chris-b1 I finally have it on Windows where it fails under a 2.7 configuration and succeeds under a 3.4 configuration.  Something to do with openpyxl writer (I think).  At least now I can compare things side by side to see what is going on. Thanks for the pointers.

**Comment by Dr-Irv** [comment]
@jreback I have figured out the issue, but am unsure what the right solution is.  Here is a table that shows the combinations that succeeded and failed, which shows the Python version, whether OpenPyxl version was specified in the requirements file, which version of OpenPyxl was used, and whether the Travis-CI run succeeded or failed

Python Version | Openpyxl version specified | Openpyxl version used| Succeeded?
--------------------|-----------------------------------|------------------------------|-----------------
2.6|None|2.2.6|YES
2.7| None | 2.2.6| YES
2.7| 1.6.2 | 1.6.2| NO
3.4|None|2.0.3|YES
3.5|None|1.8.3|NO

What is happening is that Openpyxl versions < 2.0 convert the string '40.1' to a float before writing it to the Excel file, while Openpyxl versions > 2.0 leave the string '40.1' as '40.1' .  The writer based on xlwt writes it out as a string.

These strings are created when merge_cells=false, where a '.' is used as a separator to join the values of the hierarchical index.

Possible solutions are 
a) change the openpyxl version used in the tests, (but that is a call someone else should make)
b) change the separator used when merge_cells=False to '/' or '_'
c) change the test so that the MultiIndex used are the tuples [(40,'A'),(40,'B'),(50,'A'),(50,'B')] rather than  [(40,1),(40,2),(50,1),(50,2)] 
d) Skip the test if the writer is openpyxl

Let me know which direction to go to.

**Comment by chris-b1** [comment]
@Dr-Irv - it looks like you could add in a workaround for `openpyxl < 2` by using [`set_explicit_value`](http://openpyxl.readthedocs.org/en/latest/api/openpyxl.cell.html#openpyxl.cell.cell.Cell.set_explicit_value) - basically would call that instead of `cell.value = ..` when writing the un-merged columns.

**Comment by Dr-Irv** [comment]
@jreback I think this is ready to be merged into the main branch.  Also, I included a fix for https://github.com/pydata/pandas/issues/11408 so once you do the merge, you can close that issue.

File: doc/source/whatsnew/v0.17.1.txt [hunk 5]
```diff
@@ -115,4 +115,11 @@ Bug Fixes
 - Bugs in ``to_excel`` with duplicate columns (:issue:`11007`, :issue:`10982`, :issue:`10970`)
 - Fixed a bug that prevented the construction of an empty series of dtype
   ``datetime64[ns, tz]`` (:issue:`11245`).
+
+- Bug in ``read_excel`` with multi-index containing integers (:issue:`11317`, :issue:`11328`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.17.1.txt]: only list 11317

File: pandas/core/format.py [hunk 6]
```diff
@@ -1873,8 +1873,9 @@ def _format_hierarchical_rows(self):
 
             # MultiIndex columns require an extra row
             # with index names (blank if None) for
-            # unambigous round-trip
-            if isinstance(self.columns, MultiIndex):
+            # unambigous round-trip, unless not merging,
+            # in which case the names all go on one row
```
Comment (jreback) [inline pandas/core/format.py]: add a ``#11328`` as a comment

File: pandas/io/excel.py [hunk 7]
```diff
@@ -708,7 +708,12 @@ def write_cells(self, cells, sheet_name=None, startrow=0, startcol=0):
         for cell in cells:
             colletter = get_column_letter(startcol + cell.col + 1)
             xcell = wks.cell("%s%s" % (colletter, startrow + cell.row + 1))
-            xcell.value = _conv_value(cell.val)
+            new_data_type = xcell.data_type_for_value(cell.val)
+            if (isinstance(cell.val, compat.string_types)
+                    and new_data_type != xcell.TYPE_STRING):
```
Comment (jreback) [inline pandas/io/excel.py]: you can do the ``.data_type_for_value`` inside the if (e.g. make another condition), then we ONLY do this if we have a string types

File: pandas/io/tests/test_excel.py [hunk 8]
```diff
@@ -1083,7 +1089,35 @@ def test_to_excel_multiindex(self):
             df = read_excel(reader, 'test1', index_col=[0, 1],
                               parse_dates=False)
             tm.assert_frame_equal(frame, df)
-            self.assertEqual(frame.index.names, df.index.names)
+            
+    def test_to_excel_multiindex_cols(self):
+        _skip_if_no_xlrd()
```
Comment (jreback) [inline pandas/io/tests/test_excel.py]: add the issue number here as a comment

Comment (jreback) [inline pandas/io/tests/test_excel.py]: and some commentary about what this test is doing (e.g. what the issue is)

**Comment by jreback** [comment]
pls squash when you have make corrections

**Comment by chris-b1** [comment]
I gave this another try, looks good to me!

**Comment by Dr-Irv** [comment]
@jreback I believe I've done everything you have asked, and the tests passed.

**Comment by jreback** [comment]
lgtm just need a squash 

git rebase -i master

then s everything (or use f)

then 

git push forkname branchname -f

**Comment by Dr-Irv** [comment]
@jreback OK, I think I did the squash correctly. If everything is OK, you should also close issue https://github.com/pydata/pandas/issues/11408

**Comment by jreback** [comment]
merged via https://github.com/pydata/pandas/commit/d5a04c12803f4d35d4964e6b7a4c93cb69be8331

thanks!

**Pull request closed without merging**